lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,460 @@
+			TLSConfig: tlsConfig,
+		})
+	} else {
+		redisConn = redis.NewClient(&redis.Options{
+			Addr:      connectionAddr,
+			TLSConfig: tlsConfig,
+		})
+	}
+	defer redisConn.Close()
+
+	// TODO(jakub): Currently Teleport supports only RESP2 as RESP3 is not supported by go-redis.
+	// When migration to RESP3 protocol this PING should be removed in favor of "HELLO 3" cmd.
+	// https://github.com/antirez/RESP3/blob/master/spec.md#the-hello-command-and-connection-handshake
+	pingResp := redisConn.Ping(context.Background())
+	if pingResp.Err() != nil {
+		return trace.Wrap(pingResp.Err())
+	}
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	if err := e.process(ctx, redisConn); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// readClientCmd reads commands from connected Redis client.
+func (e *Engine) readClientCmd(ctx context.Context) (*redis.Cmd, error) {
+	cmd := &redis.Cmd{}
+	if err := cmd.ReadReply(e.clientReader); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	val, ok := cmd.Val().([]interface{})
+	if !ok {
+		return nil, trace.BadParameter("failed to cast Redis value to a slice")
+	}
+
+	return redis.NewCmd(ctx, val...), nil
+}
+
+// processCmd processes commands received from connected client. Most commands are just passed to Redis instance,
+// but some require special actions:
+//  * Redis 7.0+ commands are rejected as at the moment of writing Redis 7.0 hasn't been released and go-redis doesn't support it.
+//  * RESP3 commands are rejected as Teleport currently doesn't support this version of protocol.
+//  * Subscribe related commands created a new DB connection as they change Redis request-response model to Pub/Sub.
+func (e *Engine) processCmd(ctx context.Context, redisClient redis.UniversalClient, cmd *redis.Cmd) error {
+	switch strings.ToLower(cmd.Name()) {
+	case "hello":
+		return redis.RedisError("RESP3 is not supported")
+	case "auth":
+		return e.processAuth(ctx, redisClient, cmd)
+	case "subscribe":
+		return e.subscribeCmd(ctx, redisClient.Subscribe, cmd)
+	case "psubscribe":
+		return e.subscribeCmd(ctx, redisClient.PSubscribe, cmd)
+	case "punsubscribe":
+		// TODO(jakub): go-redis doesn't expose any API for this command. Investigate alternative options.
+		return errors.New("PUNSUBSCRIBE is not supported by Teleport")
+	case "ssubscribe", "sunsubscribe":
+		// go-redis doesn't support Redis 7+ commands yet.
+		return errors.New("Redis 7.0+ is not yet supported")
+	default:
+		// Here the command is sent to the DB.
+		return redisClient.Process(ctx, cmd)
+	}
+}
+
+type redisSubscribeFn = func(ctx context.Context, channels ...string) *redis.PubSub
+
+// subscribeCmd handles subscription in Redis. A new connection is created and maintained to read published messages
+// from Redis and send them back to the client.
+func (e *Engine) subscribeCmd(ctx context.Context, subscribeFn redisSubscribeFn, cmd *redis.Cmd) error {
```

Comment:
curious how the driver handles same subscription multiple times. to the driver, it's from the same guy but we can get same subscribe from different users.

Reply:
I'm not sure if I understand the question correctly, but Redis only cares about the name of a subscription, not who is publishing messages. You can create a ACL rule to say who can publish/subscribe and who not, but that's all. It's even more interesting that messages are published on all schemas/numbers (Alice login on db 0 will still get a message published by Bob on db 2). They are also propagated across all instances in cluster mode.

Reply:
sorry. don't worry about it. I was afraid client.Subscribe will return same pubsub object or chan for the same subscription, which is not the case after i check their code.